### PR TITLE
Proj0215: Provide a compliant NuGet package icon

### PR DESCRIPTION
**Pull request by Corniel**
As suggested by Microsoft.

See #26

### CODE REVIEW

File: specs/DotNetProjectFile.Analyzers.Specs/Rules/MS_Build/NuGet packages/Define_package_info.cs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-﻿namespace Rules.MS_Build.Define_version;
+﻿namespace Rules.MS_Build.NuGet_packages;
```
Comment (LauraXiulan) [inline specs/DotNetProjectFile.Analyzers.Specs/Rules/MS_Build/NuGet packages/Define_package_info.cs:1]:
```suggestion
﻿namespace Rules.MS_Build.NuGet_packages.Define_package_info;
```

File: src/DotNetProjectFile.Analyzers/Rule.cs [hunk 2]
```diff
@@ -406,6 +406,18 @@ public static class Rule
         tags: ["Configuration", "NuGet", "package"],
         category: Category.Configuration);
 
+    public static DiagnosticDescriptor ProvideCompliantPackageIcon => New(
+        id: 0215,
+        title: "Provide a compliant NuGet package icon",
+        message: "The package icon '{0}' is {1}.",
+        description:
+            "To ensure the creation of well-formed packages, " +
+            "explicitly define the <PackageId> node or " +
```
Comment (LauraXiulan) [inline src/DotNetProjectFile.Analyzers/Rule.cs:415]: PackageId?!

File: src/DotNetProjectFile.Analyzers/IO/ImageInfo.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+                && header[0] == 0x89
+                && header[1] == 'P'
+                && header[2] == 'N'
+                && header[3] == 'G'
+                && header[4] == '\r'
+                && header[5] == '\n'
+                && header[6] == 0x1A // SUB
+                && header[7] == '\n';
+        }
+    }
```
Comment (LauraXiulan) [inline src/DotNetProjectFile.Analyzers/IO/ImageInfo.cs:42]:
```suggestion
    public static ImageInfo ReadPng(Stream image)
    {
        using var reader = new BinaryReader(image);

        if (!HasPngHeader()) return new();

        reader.BaseStream.Position = 16;

        return new()
        {
            Type = "PNG",
            Width = reader.ReadInt32BigEndian(),
            Height = reader.ReadInt32BigEndian(),
        };

        bool HasPngHeader()
        {
            var header = reader.ReadBytes(8);

            return image.Length > 16 + 13
                && header[0] == 0x89
                && header[1] == 'P'
                && header[2] == 'N'
                && header[3] == 'G'
                && header[4] == '\r'
                && header[5] == '\n'
                && header[6] == 0x1A // SUB
                && header[7] == '\n';
        }
    }
```

File: src/DotNetProjectFile.Analyzers/Analyzers/MsBuild/ProvideCompliantPackageIcon.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            .SelectMany(a => a.IncludeAndUpdate)
+            .FirstOrDefault(a => a.EndsWith(iconValue)) is { } action)
+        {
+            file = project.Path.Directory.File(action);
+        }
+
+        try
+        {
+            using var stream = file.OpenRead();
+            return ImageInfo.ReadPgn(stream);
```
Comment (LauraXiulan) [inline src/DotNetProjectFile.Analyzers/Analyzers/MsBuild/ProvideCompliantPackageIcon.cs:51]:
```suggestion
            return ImageInfo.ReadPng(stream);
```

**Review by LauraXiulan — CHANGES_REQUESTED** [review]

File: src/DotNetProjectFile.Analyzers/Rule.cs (hunk 2 above)
Comment (Corniel) [reply src/DotNetProjectFile.Analyzers/Rule.cs]: Was blindly copy-pasted. Updated now.

File: src/DotNetProjectFile.Analyzers/Rule.cs [hunk 5]
```diff
@@ -406,6 +409,16 @@ public static class Rule
         tags: ["Configuration", "NuGet", "package"],
         category: Category.Configuration);
 
+    public static DiagnosticDescriptor ProvideCompliantPackageIcon => New(
+        id: 0215,
+        title: "Provide a compliant NuGet package icon",
+        message: "The package icon '{0}' is {1}.",
+        description:
+            "To ensure the creation of well-formed packages] use an image that is " +
+            "128x128 and has a transparent background(PNG) for best viewing results.",
```
Comment (LauraXiulan) [inline src/DotNetProjectFile.Analyzers/Rule.cs:418]:
```suggestion
            "To ensure the creation of well-formed packages, use an image that is " +
            "128x128 and has a transparent background(PNG) for the best viewing results.",
```

**Review by LauraXiulan — APPROVED** [review]

File: src/DotNetProjectFile.Analyzers/Analyzers/MsBuild/ProvideCompliantPackageIcon.cs [hunk 6]
```diff
@@ -0,0 +1,58 @@
+﻿
+namespace DotNetProjectFile.Analyzers.MsBuild;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+public sealed class ProvideCompliantPackageIcon() : MsBuildProjectFileAnalyzer(Rule.ProvideCompliantPackageIcon)
```
Comment (CptWesley) [inline src/DotNetProjectFile.Analyzers/Analyzers/MsBuild/ProvideCompliantPackageIcon.cs:5]: Shouldn't this only be a requirement if the project is a package?

File: src/DotNetProjectFile.Analyzers/IO/ImageInfo.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        reader.BaseStream.Position = 16;
+
+        return new()
+        {
+            Type = "PNG",
+            Width = reader.ReadInt32BigEndian(),
+            Height = reader.ReadInt32BigEndian(),
+        };
+
+        bool HasPngHeader()
```
Comment (CptWesley) [inline src/DotNetProjectFile.Analyzers/IO/ImageInfo.cs:28]: JPG is also allowed according to the specs

File: src/DotNetProjectFile.Analyzers/IO/ImageInfo.cs [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+        if (!HasPngHeader()) return new();
+
+        reader.BaseStream.Position = 16;
+
+        return new()
+        {
+            Type = "PNG",
+            Width = reader.ReadInt32BigEndian(),
+            Height = reader.ReadInt32BigEndian(),
```
Comment (CptWesley) [inline src/DotNetProjectFile.Analyzers/IO/ImageInfo.cs:25]: Specs also specify a max size of 1MB, we can also relatively easily check this

File: src/DotNetProjectFile.Analyzers/Analyzers/MsBuild/ProvideCompliantPackageIcon.cs (hunk 6 above)
Comment (Corniel) [reply src/DotNetProjectFile.Analyzers/Analyzers/MsBuild/ProvideCompliantPackageIcon.cs:5]: Fair point. We do the same for the other package info related rules.

**Pull request merged**